### ISSUE TITLE
Request to update with new pfree patch

### ISSUE THREAD

**Issue by camprevail** [setsumi/pfreepanic64#1]
Hey mate! Thanks to Akubits, we have a much improved museca pfree edit that works like voltex, where it starts on and loops the 3rd song. Wanted to see if you could update pfreepanic64 with the new hex! :3

**Comment by setsumi** [comment]
This is a bit problematic because new edit consists of three patches instead of two. Adding another here is a mess I don't want to do honestly. On the other hand latest regular pfreepanic supports any number of patches properly but have some issue with 64-bit build that require investigation. I will try to look into it but no promises.

**Comment by camprevail** [comment]
Gotcha. I'll probably take a stab at it too. Might just make a build for myself that only uses the offset that prevents the game from ending - I would use a modded dll to have all 3 patches applied on startup.  
 
  On Mon, Sep 9, 2019 at 11:18 PM, setsumi<notifications@github.com> wrote:   
This is a bit problematic because new edit consists of three patches instead of two. Adding another here is a mess I don't want to do honestly. On the other hand latest regular pfreepanic supports any number of patches properly but have some issue with 64-bit build that require investigation. I will try to look into it but no promises.

**Comment by camprevail** [comment]
Idk what the issue you were having with the 64bit version, but I was able to update it in like 15 minutes with the three hex patches & tested on cab. Had no issues. I forked it and updated the code, but the catch is that it has to toggle between 1-song mode and Pfree mode, due to the nature of the patches. If you toggle pfree to off, it will send you to the first song since that counter is nulled instead of setting the address to two. 

TL/DR: I'll leave the issue open for now. If we can fix the patch, I'll update and make a PR. If not, we can close the issue.

**Comment by setsumi** [comment]
I assume the updated one was 'pfreepanic64' ← this one? Could you please push current changes?

About the patches, initial activation of pfree should be done as per bemanipatcher edits. Is this right? But after that I don't really understand what we do exactly to turn pfree off and on again.

**Comment by camprevail** [comment]
Yes I updated pfreepanic64, initial activation will activate all 3 patches. Deactivation only deactivates the 2nd patch. Here's the link if you want to review, I will push changes once I'm sure the hex edits are final.
https://github.com/camprevail/pfreepanic64/blob/master/Unit1.cpp

**Comment by setsumi** [comment]
Thank you for the code. Wait, isn't it already working as intended?

**Comment by camprevail** [comment]
It is, but it'd be ideal to be able to switch between Pfree and 3-song mode, instead of Pfree and 1-song mode.

**Comment by setsumi** [comment]
How? I mean, I don't play the game and can't even run it properly on dev pc (get only gray square), so the most I can do is test if memory patching is working but not the actual game behavior. Therefore I'm asking how do you imagine the procedure? Just looking at hex codes is not helping me :)

**Comment by camprevail** [comment]
:thonk: That's interesting... You could try launching with latest spice and -w for windowed mode, should work if it's windows 10. 
Anyways, idk how explain it any better than how I just did. I suck at hex edits, so Akubits made the new patches. So, I don't really know how to fix my patch, but I will ask around.

**Comment by setsumi** [comment]
Hmm..
Pfee (1 1 1) is when all 3 patches are active. This I understand, but the rest not so much.
1-song mode (1 0 1) is when you play one song more and get game over...?
3-song mode (0 0 0) is default dll without patches...?
(three digits groups above represent each of 3 patches status)
Is this right?

**Comment by camprevail** [comment]
That's all correct. But when you disable pfree on the results screen, what *should* happen is a game-over. This is currently achieved with (1-0-1). If you disable all three patches while on the results screen (0 0 0), you will go from the final song back to the 1st song, which is unwanted behavior.

**Comment by setsumi** [comment]
So the issue is with what happens after game over, i.e. unwanted 1-0-1 mode instead of desired 0-0-0 ? By the nature of current patch this is not possible without developing new one (I have zero skills in this field). But what if we make 3 modes cycle instead of toggle 2 as now? First you press button to get game over with 1-0-1 and after game is ended press again to get 0-0-0 this time (and 1-1-1 again after).

**Comment by camprevail** [comment]
Your representations are inverted, but yeah that could work. Press once to end game (1-0-1), press again to switch to 3-song mode (0-0-0), press again to enable pfree (1-1-1). Still, I find myself forgetting to press again after game-over, and by the time I load the next game, it's already on final song - too late to toggle. Some people have built crazy hooks for voltex that can detect the state of the game and reset the patches on it's own - I wish I had the knowledge to do that lol. Like the force crash dll.

**Comment by setsumi** [comment]
What about console log? If it prints something along the lines of "Entering endgame" then should be doable to monitor for that, at least in theory.

**Comment by camprevail** [comment]
Yeah Museca uses scenes, so it prints things like this: 
```[2019/09/17 20:57:09] I:Attach: in GAMEOVER
[2019/09/17 20:57:10] I:Attach: out GAMEOVER
[2019/09/17 20:57:18] I:Attach: in TITLEDEMO_FLOW
[2019/09/17 20:57:20] I:Attach: out TITLEDEMO_FLOW
```

Title demo flow is probably what we'd want to trigger the patch.

**Comment by setsumi** [comment]
Okay, I will try coding console monitor.

**Comment by setsumi** [comment]
Uploaded new code. Try testing.
P.S. Thanks for the hint. With latest spice game now runs properly windowed even on my super ancient laptop with windows 7.

**Comment by camprevail** [comment]
Not too sure what's happening, but toggling pfree off softlocks pfreepanic and museca. Toggling on works fine though.

**Comment by setsumi** [comment]
Working fine here. The thing is reading directly from console window screen buffer, therefore console window should be present and have decent size like 80x25 characters at least. Try running game windowed, when Pfree is off (1-song mode) console window should be cleared constantly.
Btw, I have added more triggers disabling 1-song mode around title menu, can upload binary if you want.

**Comment by camprevail** [comment]
What's your log level set to? 
 
  On Sat, Sep 21, 2019 at 7:26 PM, setsumi<notifications@github.com> wrote:   
Working fine here. The thing reading directly from console window screen buffer, therefore console window should be present and have decent size like 80x25 characters at least. Try running game windowed, when Pfree is off (1-song mode) console window should be cleared constantly.
Btw, I have added more triggers disabling 1-song mode around title menu, can upload binary if you want.

**Comment by setsumi** [comment]
```
	<log>
		<level __type="str">all</level>
	</log>
```
Lol. This is not good I guess.. But it shouldn't lock anything regardless

**Comment by setsumi** [comment]
Removed weird stuff. Try this new binary.

If still having softlocks try running pfreepanic64 from windows gui, not from some console/batch.

**Comment by camprevail** [comment]
Thanks, I will try that next. I tried yesterday's version again and it did work like you said, I think it was audio issues causing my crash. Restarting seems to have fixed it... Though I still noticed it caused the game to hang for a minute while it attached to the console window. 
Also, since they updated spice with a directX overlay, dx9osd can't attach and causes an exception. You're supposed to be able to disable the spice overlay with -overlaydisable, but it still causes a crash. I sent them a log and hopefully they can fix it.

**Comment by setsumi** [comment]
Well, hanging is not normal. I don't have any, there is no effect. Game runs just as smooth as before despite laptop being so weak it's running game half speed. Code just checks text inside console window 5 times per second not waiting for anything so there shouldn't be any hangs. I'm using windows 7 with admin account without UAC, starting game and app directly, spice with .bat and pfreepanic by clicking exe.

**Comment by camprevail** [comment]
Hmm okay, I will try it on my museca cab then. 
 
  On Sat, Sep 21, 2019 at 11:03 PM, setsumi<notifications@github.com> wrote:   
Well, hanging is not normal. I don't have any, there is no effect. Game runs just as smooth as before despite laptop being so weak it's running game half speed. Code just checks text inside console window 5 times per second not waiting for anything so there shouldn't be any hangs. I'm using windows 7 with admin account without UAC, starting game and app directly, spice with .bat and pfreepanic by clicking exe.

**Comment by setsumi** [comment]
Fixed silly bug.
[pfreepanic64.zip](https://github.com/setsumi/pfreepanic64/files/3639466/pfreepanic64.zip)

**Comment by setsumi** [comment]
Did it work after all?

**Comment by camprevail** [comment]
Sorry, haven't had a good chance to test yet. Will get back to you asap

**Comment by camprevail** [comment]
Finally just got a chance to test this on the cab. Works great :) Thanks! 
Could you push the latest code to the repo if you haven't already? I might tweak the OSD messages, I feel like the 1-song mode message isn't really necessary anymore. Maybe something like: End Session

**Comment by setsumi** [comment]
Oh, sorry, missed your edit. Code is latest, only binary wasn't properly uploaded to releases.

**Comment by camprevail** [comment]
Sweet.  
 
  On Tue, Nov 19, 2019 at 1:56 AM, setsumi<notifications@github.com> wrote:   
Oh, sorry, missed your edit. Code is latest, only binary wasn't properly uploaded to releases.
